Enable custom penalties for feature tweak

RaunoArike opened the pull request:
Closes #314

RaunoArike commented:
Some unrelated tests are failing for Julia 1.8 and Julia nightly, but everything related to Feature Tweak seems to work correctly

pat-alt commented on src/generators/non_gradient_based/feature_tweak/feature_tweak.jl:
```diff
@@ -7,73 +7,77 @@ mutable struct FeatureTweakGenerator <: AbstractNonGradientBasedGenerator
     end
-    return x_out
+
+    return ce
```
Since this just returns its only positional argument, should we either 1) turn this into an in-place operation `feature_tweak!(ce)` or 2) keep this, but also add an in-place version `feature_tweak!(ce)` version? 

I think I have a slight preference for option (1), because this would be a non-breaking change and it just makes sense to use the in-place version here.

RaunoArike replied:
Yes, I think it makes the most sense to just turn this into an in-place operation. Will do so in a moment.

pat-alt commented on test/generators/feature_tweak.jl:
```diff
@@ -46,18 +46,37 @@
-                                @test CounterfactualExplanations.terminated(counterfactual)
+                                objectives =
+                                    CounterfactualExplanations.Objectives.penalties_catalogue
+                                for (name, penalty) in objectives
```
Great! Does this take very long?

pat-alt reviewed (changes requested):
Just one minor change

pat-alt commented on src/generators/non_gradient_based/feature_tweak/feature_tweak.jl:
```diff
@@ -7,73 +7,77 @@ mutable struct FeatureTweakGenerator <: AbstractNonGradientBasedGenerator
+                    s′_old = ce.s′
+                    ce.s′ = reshape(es_instance, :, 1)
+                    if ce.generator.penalty(ce) < delta
+                        delta = ce.generator.penalty(ce)
```
Would it make sense to also allow for adding multiple penalties (i.e. a vector of penalty functions as for the composable generators)? Then `FeatureTweak` would fit right into the composable framework.

RaunoArike replied:
Oh right, forgot about this. I think I can add that, will check whether it works out.

RaunoArike commented:
Added the changes you requested:
- made the `feature_tweaking!()` method in-place
- removed the for-loop over the different objectives from the main test set, and added a separate test set to test the different objectives for only a single dataset and a single model. I think it makes sense to keep the Feature Tweak tests separate from the tests of the gradient-based generators. Feature Tweak only applies to Decision Tree and Random Forest models and only to binary classification problems, it would overcomplicate the tests for the gradient-based generators with if-statements if we wanted to test Feature Tweak together with them.
- I enabled using a vector of penalties, but added the `calculate_delta()` method for that instead of using the pre-existing `h()` method that we use for the gradient-based generators, since the `h()` method assumes gradient access with Zygote which we don't have for the MLJ tree models.

pat-alt reviewed (approved):
Looks good to me, great work!

pat-alt commented:
> Added the changes you requested:
> 
> * modified the `feature_tweaking!()` method to be an in-place one
> * I enabled using a vector of penalties, but added the `calculate_delta()` method for that instead of using the pre-existing `∂h()` method that we use for the gradient-based generators, since the `∂h()` method assumes gradient access with Zygote which we don't have for the MLJ tree models.
> * removed the for-loop over the different objectives from the main test set, and added a separate test set to test the different objectives for only a single dataset and a single model. I think it makes sense to keep the Feature Tweak tests separate from the tests of the gradient-based generators. Feature Tweak only applies to Decision Tree and Random Forest models and only to binary classification problems, it would overcomplicate the tests for the gradient-based generators with if-statements if we wanted to test Feature Tweak together with them.
> 
> We can still separately consider whether the test file for Feature Tweak can be made lighter, but I think some tests for different objectives are necessary: since the penalties are not calculated with the same `∂h()` function that we use for gradient-based generators, the tests should ensure that the penalty calculation logic in Feature Tweak is nevertheless compatible with all possible distance metrics.

All of that makes perfect sense, thanks!

pat-alt commented:
From my side, this can be merged
